core/src/main/java/hudson/model/Run.java

```diff
@@ -1787,6 +1789,13 @@ protected final void execute(@Nonnull RunExecution job) {
                     listener.finished(result);
                     listener.closeQuietly();
                 }
+                if (logger != null) {
```

Comment:
if `listener.finished(result)` throws an exception, the logger will not be closed